File: app/src/main/java/de/uni_marburg/mathematik/ds/serval/view/fragments/EventsFragment.java
package de.uni_marburg.mathematik.ds.serval.view.fragments;

import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import java.util.List;

import butterknife.BindView;
import de.uni_marburg.mathematik.ds.serval.R;
import de.uni_marburg.mathematik.ds.serval.controller.adapters.GenericEventAdapter;
import de.uni_marburg.mathematik.ds.serval.model.event.Event;
import de.uni_marburg.mathematik.ds.serval.model.event.EventComparator;
import de.uni_marburg.mathematik.ds.serval.model.event.GenericEvent;
import de.uni_marburg.mathematik.ds.serval.util.ImageUtil;
import de.uni_marburg.mathematik.ds.serval.util.PrefManager;
import de.uni_marburg.mathematik.ds.serval.view.activities.MainActivity;
import de.uni_marburg.mathematik.ds.serval.view.util.GridSpacingItemDecoration;

/**
 * Created by thames1990 on 28.08.17.
 */
public class EventsFragment<T extends Event> extends EventFragment<T> {
    
    private GenericEventAdapter adapter;
    
    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;
    
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }
    
    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupRecyclerView();
    }
    
    @Override
    protected int getLayout() {
        return R.layout.fragment_events;
    }
    
    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        MenuItem filterEvents = menu.findItem(R.id.action_filter_events);
        Drawable icon = ContextCompat.getDrawable(getContext(), R.drawable.filter_list);
        icon.setColorFilter(
                ContextCompat.getColor(getContext(), android.R.color.white),
                PorterDuff.Mode.SRC_IN
        );
        filterEvents.setIcon(icon);
    }
    
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_events, menu);
        PrefManager prefManager = new PrefManager(getContext());
        // Hide location filter, if the user denied location update permission
        if (!prefManager.requestLocationUpdates()) {
            menu.findItem(R.id.action_filter_events_distance).setVisible(false);
        }
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_filter_events_distance_ascending:
                adapter.filter(
                        EventComparator.DISTANCE,
                        false,
                        MainActivity.lastLocation
                );
                return true;
            case R.id.action_filter_events_distance_descending:
                adapter.filter(
                        EventComparator.DISTANCE,
                        true,
                        MainActivity.lastLocation
                );
                return true;
            case R.id.action_filter_events_measurements_ascending:
                adapter.filter(EventComparator.MEASUREMENTS, false);
                return true;
            case R.id.action_filter_events_measurements_descending:
                adapter.filter(EventComparator.MEASUREMENTS, true);
                return true;
            case R.id.action_filter_events_shuffle:
                adapter.filter(EventComparator.SHUFFLE, false);
                return true;
            case R.id.action_filter_events_time_ascending:
                adapter.filter(EventComparator.TIME, false);
                return true;
            case R.id.action_filter_events_time_descending:
                adapter.filter(EventComparator.TIME, true);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
    
    private void setupRecyclerView() {
        setupLayoutManager();
        //noinspection unchecked
        adapter = new GenericEventAdapter((List<GenericEvent>) events);
        recyclerView.setAdapter(adapter);
    }
    
    private void setupLayoutManager() {
        PrefManager prefManager = new PrefManager(getContext());
        if (prefManager.useLinearLayoutManger()) {
            recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
            recyclerView.addItemDecoration(new DividerItemDecoration(
                    getContext(),
                    DividerItemDecoration.VERTICAL
            ));
        } else if (prefManager.useGridLayoutManger()) {
            recyclerView.setLayoutManager(new GridLayoutManager(
                    getContext(),
                    prefManager.getGridLayoutManagerSpanCount()
            ));
            recyclerView.addItemDecoration(new GridSpacingItemDecoration(
                    prefManager.getGridLayoutManagerSpanCount(),
                    ImageUtil.dpToPixels(getContext(), 10),
                    true
            ));
        } else if (prefManager.useStaggeredGridLayoutManger()) {
            recyclerView.setLayoutManager(new StaggeredGridLayoutManager(
                    prefManager.getStaggeredGridLayoutManagerSpanCount(),
                    StaggeredGridLayoutManager.VERTICAL
            ));
            // TODO Add item decoration
        }
    }
}

File: app/src/main/java/de/uni_marburg/mathematik/ds/serval/view/fragments/EventsFragment.kt
package de.uni_marburg.mathematik.ds.serval.view.fragments

import android.graphics.PorterDuff
import android.os.Bundle
import android.support.v7.widget.DividerItemDecoration
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import ca.allanwang.kau.utils.*
import de.uni_marburg.mathematik.ds.serval.R
import de.uni_marburg.mathematik.ds.serval.controller.EventAdapter
import de.uni_marburg.mathematik.ds.serval.model.event.EventComparator.*
import de.uni_marburg.mathematik.ds.serval.util.Prefs
import de.uni_marburg.mathematik.ds.serval.util.consume
import de.uni_marburg.mathematik.ds.serval.util.withDividerDecoration
import de.uni_marburg.mathematik.ds.serval.view.activities.DetailActivity
import kotlinx.android.synthetic.main.fragment_events.*
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.uiThread

class EventsFragment : BaseFragment() {

    override val layout: Int
        get() = R.layout.fragment_events

    private val eventAdapter: EventAdapter by lazy {
        EventAdapter(activity!!) {
            context!!.startActivity<DetailActivity>(DetailActivity.EVENT to it)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupRecyclerView()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setupRefresh()
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater?.inflate(R.menu.menu_events, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu?) {
        menu?.let {
            super.onPrepareOptionsMenu(menu)
            val filterEvents = menu.findItem(R.id.action_filter_events)
            val icon = context!!.drawable(R.drawable.filter_list)
            icon.setColorFilter(context!!.color(android.R.color.white), PorterDuff.Mode.SRC_IN)
            filterEvents.icon = icon
        }
    }

    override fun onOptionsItemSelected(item: MenuItem) = with(eventAdapter) {
        when (item.itemId) {
            R.id.sort_distance_ascending      -> consume { sortEventsBy(DISTANCE) }
            R.id.sort_distance_descending     -> consume { sortEventsBy(DISTANCE, true) }
            R.id.sort_measurements_ascending  -> consume { sortEventsBy(MEASUREMENT) }
            R.id.sort_measurements_descending -> consume { sortEventsBy(MEASUREMENT, true) }
            R.id.sort_time_ascending          -> consume { sortEventsBy(TIME) }
            R.id.sort_time_descending         -> consume { sortEventsBy(TIME, true) }
            else                              -> super.onOptionsItemSelected(item)
        }
    }

    private fun setupRecyclerView() {
        with(context!!) {
            if (isNetworkAvailable) doAsync {
                eventAdapter.loadEvents()
                uiThread {
                    recycler_view.apply {
                        withLinearAdapter(eventAdapter)
                        withDividerDecoration(this@with, DividerItemDecoration.VERTICAL)
                        setBackgroundColor(Prefs.colorBackground)
                    }
                }
            } else toast(string(R.string.toast_network_disconnected))
        }
    }

    private fun setupRefresh() {
        with(context!!) {
            if (isNetworkAvailable) swipeRefreshLayout.apply {
                setOnRefreshListener {
                    doAsync {
                        eventAdapter.loadEvents()
                        uiThread { isRefreshing = false }
                    }
                }
            } else toast(string(R.string.toast_network_disconnected))
        }
    }
}
